aten/src/ATen/native/vulkan/VulkanAten.cpp

```diff
@@ -0,0 +1,288 @@
+#if defined(USE_VULKAN) || defined(USE_GLES)
+#include <ATen/native/vulkan/VulkanAten.h>
+#include <ATen/ATen.h>
+#include <ATen/Config.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/OpaqueTensorImpl.h>
+#include <ATen/native/UpSample.h>
+#include <ATen/native/utils/ParamUtils.h>
+
+#ifdef USE_VULKAN
+#include <ATen/native/vulkan/Vulkan.h>
+#include <ATen/native/vulkan/VulkanOps.h>
+#define VULKAN_GL vulkan
+#else
+
+#ifdef USE_GLES
+#include <ATen/native/vulkan/GLES.h>
+#define VULKAN_GL gl
+#endif
+
+#endif
+
+namespace at {
+namespace native {
+
+bool _vulkan_available() {
+  return at::native::vulkan::details::VULKAN_GL::is_available();
+}
+
+#ifdef USE_VULKAN
+using VTensor = at::native::vulkan::details::vulkan::VulkanTensor;
+#else
+
+#ifdef USE_GLES
+using VTensor = at::native::vulkan::details::gl::GLTensor;
+#endif
+
+#endif
+
+using VTensorPtr = c10::intrusive_ptr<VTensor>;
```

Comment:
now that you have shared_ptr inside VulkanTensor (i.e. it's copyable) - you don't need an extra indirection here -> just OpaqueTensorImpl<VTensor> (though you might need to do the same for GLTensor)